From 56050f27fe41e198e421aa0b306801aac3c5220a Mon Sep 17 00:00:00 2001
From: Tamizh Chelvam Raja <tamizhr@qti.qualcomm.com>
Date: Thu, 14 Aug 2025 13:01:06 +0530
Subject: [PATCH] UPSTREAM:DP: wifi: ath12k: Handle DP_RX_DECAP_TYPE_8023 type
 in Rx path

In certain case like LLC header with length field(less than 0x6000) hardware
will report a decap type as DP_RX_DECAP_TYPE_8023 in MSDU_END desc.
The packet reported in this decap type also similar to
DP_RX_DECAP_TYPE_ETHERNET2_DIX(ETH packet). So, add RX_FLAG_8023
flag to notify mac80211 that the packet is ETH type and
handle appropriately.

Patch dependency: None

Signed-off-by: Tamizh Chelvam Raja <tamizhr@qti.qualcomm.com>
---
 
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index bcf12df..501e413 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -2977,7 +2977,12 @@ static void ath12k_dp_rx_h_undecap(struct ath12k *ar, struct sk_buff *msdu,
		break;
	case DP_RX_DECAP_TYPE_8023:
-		/* TODO: Handle undecap for these formats */
+		/* Note that decap_format = 2 indicates that the decapped
+		* packet is either Ethernet 2 (DIX)  or 802.3 (uses SNAP/LLC).
+		* So, decap_format = 2 or 3 is all the same.
+		*/
+		status->flag |= RX_FLAG_8023;
+
		break;
	}
	ar->wmm_stats.total_wmm_rx_pkts[ar->wmm_stats.rx_type]++;

